pixi-multistyle-text.ts

```diff
@@ -224,94 +239,114 @@ export default class MultiStyleText extends PIXI.Text {
 		this.context.textBaseline = "alphabetic";
 		this.context.lineJoin = "round";
 
-		let basePositionY = 0;
+		let basePositionY = dropShadowPadding;
 
-		// Draw the text
+		let drawingData: TextDrawingData[] = [];
+
+		// Compute the drawing data
 		for (let i = 0; i < outputTextData.length; i++) {
 			let line = outputTextData[i];
-			let linePositionX = 0;
+			let linePositionX: number;
+
+			switch (this._style.align) {
+				case "left":
+					linePositionX = dropShadowPadding;
+					break;
+
+				case "center":
+					linePositionX = dropShadowPadding + (maxLineWidth - lineWidths[i]) / 2;
+					break;
+
+				case "right":
+					linePositionX = dropShadowPadding + maxLineWidth - lineWidths[i];
+					break;
+			}
 
 			for (let j = 0; j < line.length; j++) {
-				let textStyle = line[j].style;
-				let text = line[j].text;
-				let fontProperties = line[j].fontProperties;
+				let { style, text, fontProperties } = line[j];
 
-				this.context.font = PIXI.Text.getFontStyle(textStyle);
+				linePositionX += maxStrokeThickness / 2;
+
+				let linePositionY = maxStrokeThickness / 2 + basePositionY + fontProperties.ascent;
 
-				let strokeStyle = textStyle.stroke;
-				if (typeof strokeStyle === "number") {
-					strokeStyle = PIXI.utils.hex2string(strokeStyle);
+				if (style.valign === "bottom") {
+					linePositionY += lineHeights[i] - line[j].height - (maxStrokeThickness - style.strokeThickness) / 2;
+				} else if (style.valign === "middle") {
+					linePositionY += (lineHeights[i] - line[j].height) / 2 - (maxStrokeThickness - style.strokeThickness) / 2;
 				}
 
-				this.context.strokeStyle = strokeStyle;
-				this.context.lineWidth = textStyle.strokeThickness;
+				drawingData.push({
+					text,
+					style,
+					x: linePositionX,
+					y: linePositionY
+				});
 
-				let linePositionY = (maxStrokeThickness / 2 + basePositionY) + fontProperties.ascent;
+				linePositionX += line[j].width;
+				linePositionX -= maxStrokeThickness / 2;
+			}
 
-				if (this._style.align === "right" && linePositionX === 0) {
-					linePositionX += maxLineWidth - lineWidths[i];
-				}
-				else if (this._style.align === "center" && linePositionX === 0) {
-					linePositionX += (maxLineWidth - lineWidths[i]) / 2;
-				}
+			basePositionY += lineHeights[i];
+		}
 
-				linePositionX += maxStrokeThickness / 2;
+		this.context.save();
 
-				if (textStyle.valign === "bottom") {
-					linePositionY += lineHeights[i] - line[j].height -
-						(maxStrokeThickness - textStyle.strokeThickness) / 2;
-				} else if (textStyle.valign === "middle") {
-					linePositionY += (lineHeights[i] - line[j].height) / 2 -
-						(maxStrokeThickness - textStyle.strokeThickness) / 2;
-				}
+		// First pass: draw the shadows only
+		for (let { style, text, x, y } of drawingData) {
```

Comment:
Same here. `drawingData.forEach(({ style, text, x, y }) => { ... });` would be best